### PR TITLE
[WIP] Non-UFC Poisson demo

### PR DESCRIPTION
**Pull request by chrisrichardson**
* Create a demo for the 2D Poisson example that uses no UFC generated code
* Main change in dolfin: `FiniteElement` uses `std::function` instead of calling ufc directly

### CODE REVIEW

File: cpp/demo/poisson_noufc/main.cpp [hunk 1]
```diff
@@ -0,0 +1,591 @@
...
+//
+// The implementation is split in two files: a form file containing the
+// definition of the variational forms expressed in UFL and a C++ file
+// containing the actual solver.
+//
+// Running this demo requires the files: :download:`main.cpp`,
+// :download:`Poisson.ufl` and :download:`CMakeLists.txt`.
+//
+//
+// UFL form file
```
Comment (garth-wells) [inline cpp/demo/poisson_noufc/main.cpp]: Best to delete doc that is no longer relevant.

File: cpp/demo/poisson_noufc/main.cpp [hunk 2]
```diff
@@ -0,0 +1,591 @@
...
+
+using namespace dolfin;
+
+// Then follows the definition of the coefficient functions (for
+// :math:`f` and :math:`g`), which are derived from the
+// :cpp:class:`Expression` class in DOLFIN
+//
+// .. code-block:: cpp
+
+class P1Element
```
Comment (garth-wells) [inline cpp/demo/poisson_noufc/main.cpp]: I don't see the need for a class. Could use a namespace if you want to group functions.

File: cpp/demo/poisson_noufc/main.cpp (hunk 1 above)
Comment (chrisrichardson) [reply cpp/demo/poisson_noufc/main.cpp]: :+1:

**Comment by chrisrichardson** [comment]
I'm now slightly holding off merging, as I am working on interface changes to `tabulate_tensor`...

**Comment by garth-wells** [comment]
@chrisrichardson Is this ready now? If not, maybe you could change it to a work-in-progress PR? (There is an option for that now.)

**Comment by chrisrichardson** [comment]
How do I change to WIP?

**Pull request closed without merging**